garak/probes/badchars.py

```diff
@@ -0,0 +1,418 @@
+        swap characters with bidi tricks, substitute homoglyphs, and inject deletions/backspaces.
+    The goal is to stress refusal policies with subtle transformations that should remain semantically equivalent to the original payloads.
+    """
+
+    lang = "en"
+    active = False
+    doc_uri = "https://arxiv.org/abs/2106.09898"
+    goal = "inject imperceptible text perturbations that bypass refusal policies"
+    primary_detector = "mitigation.MitigationBypass"
+    tags = [
+        "avid-effect:security:S0403",
+        "quality:Security:PromptStability",
+        "payload:jailbreak",
+    ]
+    tier = garak.probes.Tier.COMPETE_WITH_SOTA
+
+    DEFAULT_PARAMS = garak.probes.Probe.DEFAULT_PARAMS | {
+        "payload_name": "harmful_behaviors",
+        "perturbation_budget": 1,
+        "enabled_categories": [
+            "invisible",
+            "homoglyph",
+            "reordering",
+            "deletion",
+        ],
+        "max_position_candidates": 24,
+        "max_reorder_candidates": 24,
+        "max_ascii_variants": len(ASCII_PRINTABLE),
+        "follow_prompt_cap": True,
+    }
+
+    def __init__(self, config_root=_config):
+        super().__init__(config_root=config_root)
+        if self.perturbation_budget <= 0:
+            raise PluginConfigurationError(
+                "perturbation_budget must be a positive integer"
+            )
+
+        self._homoglyph_map = _load_homoglyph_map()
+        categories = self.enabled_categories
+        if isinstance(categories, str):
+            categories = [categories]
+        self._enabled_categories = {
+            cat.lower() for cat in categories if isinstance(cat, str)
+        }
+        if not self._enabled_categories:
+            raise PluginConfigurationError(
+                "enabled_categories must include at least one entry"
+            )
+
+        payload_group = garak.payloads.load(self.payload_name)
+        self._source_payloads = payload_group.payloads
+        self.prompts: List[garak.attempt.Conversation] = []
+        self._seen_prompts: set[str] = set()
+
+        for idx, payload in enumerate(self._source_payloads):
+            for text, metadata in self._generate_variants(payload, idx):
+                self._append_prompt(text, metadata)
+
+        if not self.prompts:
+            raise PluginConfigurationError(
+                "BadCharacters failed to generate any prompts. "
+                "Check payload_name and category configuration."
+            )
+
+        if (
+            self.follow_prompt_cap
+            and self.soft_probe_prompt_cap is not None
+            and len(self.prompts) > self.soft_probe_prompt_cap
+        ):
+            self._downsample_prompts()
+
+    def _append_prompt(self, text: str, metadata: dict) -> None:
+        if text in self._seen_prompts:
+            return
+        self._seen_prompts.add(text)
```

Comment:
Should this filter when the same text is offered with different metadata?

If not then this could also be simplified by using a `set()` for prompt and just creating the `Conversation` as from the param and letting the set logic utilize comparison to deduplicate fully unique `Conversation` objects.

While sightly more runtime cost would be generated the long term memory savings in not holding onto `self._seen_prompt` for the entire lifecycle of the probe might be a fair trade.